### PR TITLE
Last call changes

### PR DESCRIPTION
**Pull request by nightwatchcyber**
- Added language and example regarding URI encoding (#176)
- Add "Expires" field (#181)
- Changed language from "directive" to "field" (#182)
- Addressing last call feedback (#175, #179 and #180)
- Clarifying order of fields (#174)

### CODE REVIEW

File: draft-foudil-securitytxt.md [hunk 1]
```diff
@@ -35,8 +35,8 @@ When security vulnerabilities are discovered by independent
 security researchers, they
 often lack the channels to report them properly. As a result,
 security vulnerabilities may be left unreported. This document defines a format
-("security.txt") to help organizations describe the process for security
-researchers to follow in order to report security vulnerabilities.
+("security.txt") to help organizations describe their vulnerability disclosure practices
+in order to make it easier for researchers to report security vulnerabilities.
```
Comment (EdOverflow) [inline draft-foudil-securitytxt.md]:
```suggestion
to make it easier for researchers to report security vulnerabilities.
```

File: draft-foudil-securitytxt.md [hunk 2]
```diff
@@ -46,33 +46,35 @@ researchers to follow in order to report security vulnerabilities.
...
-policies and contact information for organizations in order to report
+these address the issue of how security researchers can locate vulnerability disclosure
+practices and contact information for organizations in order to report
 security vulnerabilities.
 
-In this document, we define a richer, machine-parsable and extensible way
+In this document, we define a richer and extensible way
 for organizations to communicate information about their security disclosure
-policies, which is not limited to email and also allows for additional features
+practices and ways to contact them, which is not limited to email and also allows for additional features
```
Comment (EdOverflow) [inline draft-foudil-securitytxt.md]: This sentence is too wordy.

File: draft-foudil-securitytxt.md [hunk 3]
```diff
@@ -46,33 +46,35 @@ researchers to follow in order to report security vulnerabilities.
...
 such as encryption. This format is designed to help
 assist with the security disclosure process by making it easier
 for organizations to designate the preferred steps for researchers to take
-when trying to reach out to them with security vulnerabilities.
+when trying to reach out to them with security vulnerabilities as well as provide
+additional information about their vulnerability disclosure practices useful to security researchers.
```
Comment (EdOverflow) [inline draft-foudil-securitytxt.md]: This sentence is too wordy. Needs rephrasing.

File: draft-foudil-securitytxt.md [hunk 4]
```diff
@@ -96,35 +98,37 @@ Development of this draft takes place on Github at: https://github.com/securityt
 # The Specification
 
 This document defines a text file to be placed in a known location
-that provides information for security researchers to assist
-in disclosing security vulnerabilities.
+that provides information for security researchers about the vulnerability disclosure
+practices of a particular organization in order to assist them in disclosing security vulnerabilities.
```
Comment (EdOverflow) [inline draft-foudil-securitytxt.md]:
```suggestion
practices of a particular organization to assist them in disclosing security vulnerabilities.
```

File: draft-foudil-securitytxt.md [hunk 5]
```diff
@@ -285,10 +284,22 @@ Example of an OpenPGP key being referenced by its fingerprint:
...
+### Expires {#expires}
+
+This field indicates the date/time after which the data contained in the "security.txt"
+file is considered stale and should not be used.
+The value of this field follows the format defined in section 3.3 of {{!RFC5322}}.
+
+This field MUST NOT appear more than once.
+
+~~~~~~~~~~
+Expires: Thu, 31 Dec 2020 18:37:07 -0800
```
Comment (EdOverflow) [inline draft-foudil-securitytxt.md]: Is the date arbitrary? Is it the implementer's sole discretion to set a realistic expiration date? How does one determine what a suitable expiration date is?

File: draft-foudil-securitytxt.md [hunk 6]
```diff
@@ -297,32 +308,32 @@ Hiring: https://example.com/jobs.html
 
 ### Policy {#policy}
 
-This directive indicates a link to where the security policy and/or
-disclosure policy is located. This can help security researchers understand
-what an organization is looking for and how to report security vulnerabilities.
-If this directive indicates a web URL, then it MUST begin with "https://"
+This field indicates a link to where the vulnerability disclosure policy is located.
```
Comment (EdOverflow) [inline draft-foudil-securitytxt.md]:
```suggestion
This field indicates a link to where the vulnerability-disclosure policy is located.
```

File: draft-foudil-securitytxt.md [hunk 7]
```diff
@@ -297,32 +308,32 @@ Hiring: https://example.com/jobs.html
...
+This can help security researchers understand
+the organization's vulnerability reporting practices.
```
Comment (EdOverflow) [inline draft-foudil-securitytxt.md]:
```suggestion
the organization's vulnerability-reporting practices.
```

File: draft-foudil-securitytxt.md [hunk 8]
```diff
@@ -377,17 +388,14 @@ Version: GnuPG v2.2
...
+Web-based services MUST place the security.txt file under the "/.well-known/" path; e.g. https://example.com/.well-known/security.txt
 as per {{!RFC8615}}. For legacy compatibility, a security.txt file might be placed at the top-level path
-or redirect (as per section 6.4 of {{!RFC7231}}) to the security.txt file under the /.well-known/ path.
-
-If retrieval of a "security.txt" file from the top-level path results in a redirect (as per
-section 6.4 of {{!RFC7231}}), the implementors MUST NOT follow that
-redirect if it leads to another domain or subdomain
-but SHOULD follow that redirect within the same domain name
-(but not different subdomain on the same domain).
+or redirect (as per section 6.4 of {{!RFC7231}}) to the security.txt file under the /.well-known/ path. If a "security.txt" file
```
Comment (EdOverflow) [inline draft-foudil-securitytxt.md]:
```suggestion
or redirect (as per section 6.4 of {{!RFC7231}}) to the security.txt file under the "/.well-known/" path. If a "security.txt" file
```

File: draft-foudil-securitytxt.md [hunk 9]
```diff
@@ -550,42 +559,46 @@ organization operating that website to be a way to signal to researchers
 that permission to test that particular site or project is denied. This might result in pushback against
 researchers reporting security issues to that organization.
 
-Therefore, implementors MUST NOT assume that presence or absence of
+Therefore, implementors shouldn't assume that presence or absence of
 a "security.txt" file grants or denies permission for security testing.
-Any such permission MAY be defined in a security or disclosure policy
-(as per {{policy}}) or a new directive (as per {{extensibility}}).
+Any such permission may be indicates in the company's vulnerability disclosure policy
```
Comment (EdOverflow) [inline draft-foudil-securitytxt.md]:
```suggestion
Any such permission may be indicated in the company's vulnerability-disclosure policy
```

File: draft-foudil-securitytxt.md [hunk 10]
```diff
@@ -550,42 +559,46 @@ organization operating that website to be a way to signal to researchers
...
+As an additional layer of protection, it is also recommended that
 organizations digitally sign their "security.txt" file with OpenPGP (as per {{signature}}).
 Also, to protect security reports from being tampered with or observed while in transit,
-organizations SHOULD specify encryption keys (as per {{encryption}}) unless
+organizations should specify encryption keys (as per {{encryption}}) unless
 HTTPS is being used.
 
 However, the determination of validity of such keys is out of scope
-for this specification. Implementors MUST establish other secure means to
+for this specification. Researches need to establish other secure means to
```
Comment (EdOverflow) [inline draft-foudil-securitytxt.md]:
```suggestion
for this specification. Security researches need to establish other secure means to
```

File: draft-foudil-securitytxt.md [hunk 11]
```diff
@@ -594,15 +607,16 @@ Similar to concerns in {{!RFC2142}}, denial of service attacks via spam reports
...
+Organizations need to weigh the advantages of publishing this file versus
 the possible disadvantages and increased resources required to triage
 security reports.
 
-Security researchers SHOULD consult the organization's policy, if available,
-before submitting reports in an automated fashion or as resulting from
-automated scans.
+Security researchers should consult the organization's vulnerability disclosure policy, if available,
+and review the contact information and/or resources referenced within the "security.txt"
+file before submitting reports in an automated fashion or as resulting from automated scans.
```
Comment (EdOverflow) [inline draft-foudil-securitytxt.md]: Too wordy. Needs rephrasing.

File: draft-foudil-securitytxt.md (hunk 5 above)
Comment (nightwatchcyber) [reply draft-foudil-securitytxt.md]: I added a reference to the staleness section from the security considerations section

File: draft-foudil-securitytxt.md (hunk 1 above)
Comment (nightwatchcyber) [reply draft-foudil-securitytxt.md]: will change

File: draft-foudil-securitytxt.md (hunk 2 above)
Comment (nightwatchcyber) [reply draft-foudil-securitytxt.md]: Rewording this - being that we already mentioned this is intended for researchers to be consumed, I took that out and only left the part about the organizations.

File: draft-foudil-securitytxt.md (hunk 3 above)
Comment (nightwatchcyber) [reply draft-foudil-securitytxt.md]: will do

File: draft-foudil-securitytxt.md (hunk 4 above)
Comment (nightwatchcyber) [reply draft-foudil-securitytxt.md]: will do

File: draft-foudil-securitytxt.md (hunk 6 above)
Comment (nightwatchcyber) [reply draft-foudil-securitytxt.md]: not sure why a dash would work here?

File: draft-foudil-securitytxt.md (hunk 7 above)
Comment (nightwatchcyber) [reply draft-foudil-securitytxt.md]: same question about the dash as above

File: draft-foudil-securitytxt.md (hunk 8 above)
Comment (nightwatchcyber) [reply draft-foudil-securitytxt.md]: changed

File: draft-foudil-securitytxt.md (hunk 9 above)
Comment (nightwatchcyber) [reply draft-foudil-securitytxt.md]: will do

File: draft-foudil-securitytxt.md (hunk 10 above)
Comment (nightwatchcyber) [reply draft-foudil-securitytxt.md]: will change

File: draft-foudil-securitytxt.md (hunk 11 above)
Comment (nightwatchcyber) [reply draft-foudil-securitytxt.md]: will do

File: draft-foudil-securitytxt.md (hunk 10 above)
Comment (EdOverflow) [reply draft-foudil-securitytxt.md]: Actually, there is a typo there. It should be:

```suggestion
for this specification. Security researchers need to establish other secure means to
```

Comment (nightwatchcyber) [reply draft-foudil-securitytxt.md]: I fixed it

File: draft-foudil-securitytxt.md [hunk 12]
```diff
@@ -499,31 +499,31 @@ the "security.txt" file as well or setup a redirect to their own site.
...
-files (as per {{signature}}), use the canonical directive to sign the location
+To protect against this, organizations should digitally sign their "security.txt"
+files (as per {{signature}}), use the "Canonical" field to sign the locations
 of the file (as per {{canonical}}), and regularly monitor the file and
 the referenced resources to detect tampering.
 
-Security researchers SHOULD check the "security.txt" file including verifying
+Security researchers should triage the "security.txt" file including verifying
 the digital signature and checking any available historical records before using the information
 contained in the file. If "security.txt" file looks suspicious or compromised,
```
Comment (joker314) [inline draft-foudil-securitytxt.md]: (Not introduced by this PR, but probably good to fix this typo). One of--
```suggestion
 contained in the file. If a "security.txt" file looks suspicious or compromised,
```
```suggestion
 contained in the file. If the "security.txt" file looks suspicious or compromised,
```

File: draft-foudil-securitytxt.md [hunk 13]
```diff
@@ -96,35 +91,37 @@ Development of this draft takes place on Github at: https://github.com/securityt
...
-For web-based services, the file MUST be accessible via the Hypertext Transfer Protocol (HTTP) {{!RFC1945}} as a resource of Internet Media Type "text/plain" with the default charset parameter set to "utf-8" per section 4.1.3 of {{!RFC2046}}, and it MUST be served with "https" (as per section 2.7.2 of {{!RFC7230}}). For file systems and version control repositories a "security.txt" file SHOULD be placed in the root directory of a particular file system or source code project.
+For web-based services, the file MUST be accessible via the Hypertext Transfer Protocol (HTTP) 1.0 {{!RFC1945}} or higher version, as a resource of Internet Media Type "text/plain" with the default charset parameter set to "utf-8" per section 4.1.3 of {{!RFC2046}}, and it MUST be served with "https" (as per section 2.7.2 of {{!RFC7230}}). For file systems a "security.txt" file SHOULD be placed in the root directory of a particular file system.
 
-This text file contains multiple directives
-with different values. The "directive" is the first part of a field all the way up
+This text file contains multiple fields with different values. A field contains a name which is the first part of a field all the way up
```
[the remark this answers is not in the document]
Comment (nightwatchcyber) [reply draft-foudil-securitytxt.md]: will do

File: draft-foudil-securitytxt.md [hunk 14]
```diff
@@ -438,16 +433,17 @@ sign-header      =  < headers and line from section 7 of [RFC4880] >
...
                     *line [lang-field eol] *line
-                    ; the order of elements is not important
+                    ; the order of fields within the file is not important
 
 line             =  (field / comment) eol
 
 eol              =  *WSP [CR] LF
 
 field            =  ack-field /
+                    can-field /
```
[the remark this answers is not in the document]
Comment (nightwatchcyber) [reply draft-foudil-securitytxt.md]: As per IETF feedback, this was changed so Canonical can appear multiple times. This was done to address the use case where multiple redirects point to the same file.

File: draft-foudil-securitytxt.md (hunk 12 above)
Comment (nightwatchcyber) [reply draft-foudil-securitytxt.md]: will do

File: draft-foudil-securitytxt.md [hunk 15]
```diff
@@ -94,11 +94,16 @@ This document defines a text file to be placed in a known location
...
-The file is named "security.txt", and this file MUST be placed under the
-/.well-known/ path ("/.well-known/security.txt") {{!RFC8615}} of a domain name or IP address for web
-properties. For legacy compatibility, a security.txt file might be placed at the top level path (see {{weblocation}}).
+By convention, the file is named "security.txt".
 
-For web-based services, the file MUST be accessible via the Hypertext Transfer Protocol (HTTP) 1.0 {{!RFC1945}} or higher version, as a resource of Internet Media Type "text/plain" with the default charset parameter set to "utf-8" per section 4.1.3 of {{!RFC2046}}, and it MUST be served with "https" (as per section 2.7.2 of {{!RFC7230}}). For file systems a "security.txt" file SHOULD be placed in the root directory of a particular file system.
+When made available on HTTP servers, it MUST be placed under the
+/.well-known/ path (as "/.well-known/security.txt") {{!RFC8615}} of a domain name or IP address.
+For legacy compatibility, a security.txt file might be placed at the top level path (see {{weblocation}}).
+For file systems a "security.txt" file SHOULD be placed in the root directory of a particular file system.
```
Comment (EdOverflow) [inline draft-foudil-securitytxt.md]:
```suggestion
For file systems a "security.txt" file SHOULD be placed in the root directory of the file system.
```

I do not think "particular" is needed here.

File: draft-foudil-securitytxt.md [hunk 16]
```diff
@@ -94,11 +94,16 @@ This document defines a text file to be placed in a known location
...
+
+On HTTP servers, the file MUST be accessed via the Hypertext Transfer Protocol (HTTP) 1.0 {{!RFC1945}} or higher version
```
Comment (EdOverflow) [inline draft-foudil-securitytxt.md]:
```suggestion
On HTTP servers, the file MUST be accessed via the Hypertext Transfer Protocol (HTTP) 1.0 {{!RFC1945}} or a higher version
```

File: draft-foudil-securitytxt.md [hunk 17]
```diff
@@ -94,11 +94,16 @@ This document defines a text file to be placed in a known location
...
+and the "https" scheme (as per section 2.7.2 of {{!RFC7230}}). It MUST have a Content-Type of "text/plain"
```
Comment (EdOverflow) [inline draft-foudil-securitytxt.md]: > On HTTP servers, the file MUST be accessed via the Hypertext Transfer Protocol 1.0 or a higher version and the "https" scheme.

This sentence is quite difficult to follow. Especially with all the inline references. Might need rephrasing.

File: draft-foudil-securitytxt.md [hunk 18]
```diff
@@ -122,8 +127,8 @@ contain URIs using percent-encoding (as per section 2.1 of {{!RFC3986}}).
 A "security.txt" file MUST only apply to the domain or IP address in the URI used to retrieve it,
 not to any of its subdomains or parent domains. A "security.txt" file that is found
 in a file system MUST only apply to the folder
-or repository in which it is located, and not to any of its parent or sibling folders,
-or repositories. However, it will apply to all subfolders.
+in which it is located, and not to any of its parent or sibling folders.
+However, it will apply to all subfolders.
```
Comment (EdOverflow) [inline draft-foudil-securitytxt.md]: I would reword this as follows:

> A "security.txt" file that is found in a file system MUST only apply to the folder in which it is located and that folder's subfolders. The file does not apply to any of the folder's parent or sibling folders.

File: draft-foudil-securitytxt.md (hunk 15 above)
Comment (nightwatchcyber) [reply draft-foudil-securitytxt.md]: done

File: draft-foudil-securitytxt.md (hunk 16 above)
Comment (nightwatchcyber) [reply draft-foudil-securitytxt.md]: done

File: draft-foudil-securitytxt.md (hunk 17 above)
Comment (nightwatchcyber) [reply draft-foudil-securitytxt.md]: reworded

File: draft-foudil-securitytxt.md (hunk 18 above)
Comment (nightwatchcyber) [reply draft-foudil-securitytxt.md]: done

**Pull request merged**